nncf/pruning/pruning_node_selector.py

```diff
@@ -0,0 +1,203 @@
+"""
+ Copyright (c) 2021 Intel Corporation
+ Licensed under the Apache License, Version 2.0 (the "License");
+ you may not use this file except in compliance with the License.
+ You may obtain a copy of the License at
+      http://www.apache.org/licenses/LICENSE-2.0
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+"""
+
+from typing import Dict, List, Optional
+
+from nncf.pruning.utils import get_sources_of_node, is_depthwise_conv, get_previous_conv, get_first_pruned_nodes, \
+    get_last_pruned_nodes, is_grouped_conv, is_conv_with_downsampling
+from nncf.pruning.export_helpers import PruningOperationsMetatypeRegistry
+from nncf.pruning.model_analysis import ModelAnalyzer, Clusterization, cluster_special_ops, NodesCluster
+from nncf.dynamic_graph.graph import NNCFGraph, NNCFNode
+from nncf.nncf_logger import logger as nncf_logger
+from nncf.utils import should_consider_scope
+
+
+class PruningNodeSelector:
+    """
+        Determines which of the nodes with pruning types should be pruned
+        and which of them should be pruned together
+    """
+    def __init__(self,
+                 pruning_operator_metatypes: PruningOperationsMetatypeRegistry,
+                 prune_operations: List[str],
+                 grouping_operations: List[str],
+                 ignored_scopes: Optional[List[str]],
+                 target_scopes: Optional[List[str]],
+                 ignore_frozen_layers: bool,
+                 prune_first: bool,
+                 prune_last: bool,
+                 prune_downsample_convs: bool):
+        """
+        :param pruning_operator_metatypes: registry with operation metatypes pruning algorithm is aware of
```

Comment:
I think that the algorithm's logic is tightly coupled to *all* the operations in the PRUNING_OPERATOR_METATYPES, since there are also quite a lot of auxiliary operations being registered in PRUNING_OPERATOR_METATYPES. Is there really a need to pass it as an argument to the object? Will there be a situation when we will need to create objects with different values of `pruning_operator_metatypes` and this will lead to different, valid behaviors?

Reply:
My point is that the majority (perhaps even all) of the contents of the registry itself really isn't something configurable and variable across use cases. Rather, these are required to abstract away the algorithm logic from the actual operator or module names, and, hopefully, frameworks. The registry is a property of the pruning algorithm as a whole.

Reply:
The idea is to use different registries for different frameworks. For each framework there will be its own registry based on framework-specific classes (all inherent from DefaultMetaOp) and the same set of registry keys. So, one of these registries will be passed to framework-agnostic PruningNodeSelector.

Reply:
Extended `pruning_operator_metatypes` description.